substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java

```diff
@@ -118,6 +119,10 @@ public final class HeapImpl extends Heap {
     /** A cached list of all the classes, if someone asks for it. */
     private List<Class<?>> classList;
 
+    /** A cached list of all the loaded classes, if someone asks for it. */
+    private List<Class<?>> loadedClasses;
```

Comment:
If the cache is necessary, you should also consider the impact of this cache from a memory consumption perspective. To assess that you would need to run your experiments before/after and compare max RSS consumption.

Reply:
That's a great idea! I did just that and found some very interesting results.

Experiment routine: Use the same hibernate-orm Quarkus quickstart that I used in the previous two experiments. Use use the `ps` command to obtain the RSS as described in the [Quarkus docs](https://quarkus.io/guides/native-reference#rss). Repeat 5 times with/without the optimization changes.

Results:
For both with/without the optimization changes, the first RSS measurement taken after starting the app is not statistically different (between 101000 - 110000 with much variance in between). For comparison, running the app in java mode results in RSS 200000+.   However, the interesting result is that, in the unoptimized version, RSS steadily creeps up a total of about 7000 over a 10s interval (with no activity hitting the app endpoints). I believe this is due to the JFR periodic task allocating new TLABs every second. The optimized version does not have this problem.